shared/lsif/ranges.ts

```diff
@@ -0,0 +1,350 @@
+import * as sourcegraph from 'sourcegraph'
+import gql from 'tagged-template-noop'
+import { queryGraphQL as sgQueryGraphQL, QueryGraphQLFn } from '../util/graphql'
+import { nodeToLocation, LocationConnectionNode } from './locations'
+import { HoverPayload } from './hover'
+import {
+    GenericLSIFResponse,
+    queryLSIF,
+    lsifRequest,
+    rangeFragment,
+    simpleResourceFragment,
+    markdownFragment,
+} from './api'
+
+/** The size of the bounds on each ranges request. */
+const RANGE_WINDOW_SIZE = 100
+
+/**
+ * The maximum number of documents to store windows for. This value can be
+ * relatively low as only the current document really matters. We just want
+ * to keep a handful of values around in case the user navigates back to a
+ * previous page.
+ *
+ * Note: Alternatively, we may want to enforce a maximum number of windows
+ * that can remain resident in the extension at one time as well. This will
+ * make it easier to browse much larger files, but this seems like it is a
+ * good enough v1 to stop things from slowing to a crawl in the vast majority
+ * of circumstances.
+ */
+const WINDOW_CACHE_CAPACITY = 5
+
+/** The type returned by makeRangeWindowFactory. */
+export type RangeWindowFactoryFn = (
+    doc: sourcegraph.TextDocument,
+    position: sourcegraph.Position
+) => Promise<CodeIntelligenceRange | null>
+
+/** A range and a subset of its intelligence data. */
+export interface CodeIntelligenceRange {
+    range: sourcegraph.Range
+    definitions?: sourcegraph.Location[]
+    references?: sourcegraph.Location[]
+    hover?: HoverPayload
+}
+
+/** A set of code intelligence ranges and the line bounds in which they are contained. */
+interface RangeWindow {
+    startLine: number
+    endLine: number
+    ranges: Promise<CodeIntelligenceRange[] | null>
+}
+
+/**
+ * Create a factory function that returns code intelligence ranges for the given
+ * document and position. This will request bulk data from the GraphQL API
+ * (a range around the given position) and cache the result so that similar queries
+ * will not have to make a subsequent network request.
+ *
+ * The data returned from this function is precise but not complete. Notably, it is
+ * missing any definitions and references that do not exist in the same bundle. In
+ * order to get cross-repository and cross-root intelligence, the provider must fall
+ * back to an explicit request for that range when the data here is not sufficient.
+ *
+ * @param queryGraphQL The function used to query the GraphQL API.
+ */
+export async function makeRangeWindowFactory(
+    queryGraphQL: QueryGraphQLFn<any> = sgQueryGraphQL
+): Promise<RangeWindowFactoryFn> {
+    if (!(await hasRangesQuery(queryGraphQL))) {
+        // No-op if the instance doesn't support bulk loading
+        return () => Promise.resolve(null)
+    }
+
+    const cache = new Map<
+        sourcegraph.TextDocument,
```

Comment:
You may wanna store by the URI string just in case a text document is the same URI but not referentially equal (they _should_ generally be, unless a text document gets removed, and then added later again)